python/descriptor_containers.c

```diff
@@ -0,0 +1,670 @@
+  PyUpb_ModuleState *s = PyUpb_ModuleState_Get();
+  PyUpb_DescriptorIterator *iter =
+      (void *)PyType_GenericAlloc(s->descriptor_iterator_type, 0);
+  iter->funcs = funcs;
+  iter->parent = parent;
+  iter->parent_obj = parent_obj;
+  iter->index = 0;
+  Py_INCREF(iter->parent_obj);
+  return &iter->ob_base;
+}
+
+PyObject* PyUpb_DescriptorIterator_IterNext(PyObject* _self) {
+  PyUpb_DescriptorIterator *self = (PyUpb_DescriptorIterator*)_self;
+  int size = self->funcs->get_elem_count(self->parent);
+  if (self->index >= size) return NULL;
+  const void *elem = self->funcs->index(self->parent, self->index);
+  self->index++;
+  return self->funcs->get_elem_wrapper(elem);
+}
+
+static PyType_Slot PyUpb_DescriptorIterator_Slots[] = {
+  {Py_tp_dealloc, PyUpb_DescriptorIterator_Dealloc},
+  {Py_tp_iter, PyObject_SelfIter},
+  {Py_tp_iternext, PyUpb_DescriptorIterator_IterNext},
+  {0, NULL}
+};
+
+static PyType_Spec PyUpb_DescriptorIterator_Spec = {
+  PYUPB_MODULE_NAME ".DescriptorIterator",   // tp_name
+  sizeof(PyUpb_DescriptorIterator),          // tp_basicsize
+  0,                                         // tp_itemsize
+  Py_TPFLAGS_DEFAULT,                        // tp_flags
+  PyUpb_DescriptorIterator_Slots,
+};
+
+// -----------------------------------------------------------------------------
+// GenericSequence
+// -----------------------------------------------------------------------------
+
+typedef struct {
+  PyObject_HEAD
+  const PyUpb_GenericSequence_Funcs *funcs;
+  const void *parent;    // upb_msgdef*, upb_symtab*, etc.
+  PyObject *parent_obj;  // Python object that keeps parent alive, we own a ref.
+} PyUpb_GenericSequence;
+
+PyObject *PyUpb_GenericSequence_New(
+    const PyUpb_GenericSequence_Funcs *funcs, const void *parent,
+    PyObject *parent_obj) {
+  PyUpb_ModuleState *s = PyUpb_ModuleState_Get();
+  PyUpb_GenericSequence *seq =
+      (PyUpb_GenericSequence *)PyType_GenericAlloc(s->generic_sequence_type, 0);
+  seq->funcs = funcs;
+  seq->parent = parent;
+  seq->parent_obj = parent_obj;
+  Py_INCREF(parent_obj);
+  return &seq->ob_base;
+}
+
+PyUpb_GenericSequence *PyUpb_GenericSequence_Get(PyObject *obj) {
+  assert(Py_TYPE(obj) == PyUpb_ModuleState_Get()->generic_sequence_type);
+  return (PyUpb_GenericSequence*)obj;
+}
+
+static void PyUpb_GenericSequence_Dealloc(PyObject *_self) {
+  PyUpb_GenericSequence *self = PyUpb_GenericSequence_Get(_self);
+  Py_CLEAR(self->parent_obj);
+  PyUpb_Dealloc(self);
+}
+
+static Py_ssize_t PyUpb_GenericSequence_Length(PyObject* _self) {
+  PyUpb_GenericSequence *self = PyUpb_GenericSequence_Get(_self);
+  return self->funcs->get_elem_count(self->parent);
+}
+
+static PyObject *PyUpb_GenericSequence_GetItem(PyObject *_self,
+                                               Py_ssize_t index) {
+  PyUpb_GenericSequence *self = PyUpb_GenericSequence_Get(_self);
+  Py_ssize_t size = self->funcs->get_elem_count(self->parent);
+  if (index < 0 || index >= size) {
```

Comment:
In this case I'm not sure I agree.

As currently written, I read this as "if index is too small, or if it is too big." Index is consistently on the left because it's the variable whose validity is in question.

With your rewrite, it looks like we are first testing the validity of index, and then testing the validity of size against index. But size is known-good. For me personally your rewritten version is harder to visually scan.

Reply:
I read it as "index is outside the range [0, size)" but if you feel otherwise feel free to leave as is.